manifests/candlepin.pp

```diff
@@ -130,6 +115,29 @@
       key_mode    => '0440',
     }
 
+    file { $keystore_password_path:
+      ensure  => file,
+      content => $keystore_password,
+      owner   => 'root',
+      group   => $group,
+      mode    => '0440',
+    }
+
+    keystore { "${keystore}:tomcat":
+      ensure        => 'present',
+      certificate   => $tomcat_cert,
+      private_key   => $tomcat_key,
+      ca_file       => $ca_cert,
+      password_file => $keystore_password_path,
+    }
+
+    file { $keystore:
+      ensure => file,
+      owner  => 'root',
+      group  => $group,
+      mode   => '0640',
+    }
```

Comment:
Interesting question because it made me think. Shouldn't this type be named `keystore_certificate`? There can be multiple certificates in a single keystore.

If there is a `keystore` type that only ensures a keystore exists at a certain place, managing the permissions on that makes sense. However, with `keystore_certificate` there may be multiple instances pointing to the same keystore and you may end up with conflicting definitions.

Reply:
The artemis access may be one example. I can imagine that once we have a layout of:
* `candlepin.example.com`
* `foreman01.example.com`
* `foreman02.example.com`

Then `foreman01` and `foreman02` might have separate client certificates. In the current layout, we'd need to import both public parts on `candlepin.example.com` to trust, right?

Reply:
There is a `keystore` and a `truststore` present.

The `keystore` contains the server keypair for Tomcat and the Artemis broker running "inside" of Tomcat. Both configurations, server.xml for Tomcat and broker.xml for Artemis, use the keystore directly and expect to contain a server keypair.

The `truststore` contains the client certificate(s) for any clients connecting to Tomcat and Artemis.

So the `truststore` may end up with multiple client certificates, but the keystore should only need a single server keypair unless Tomcat and Artemis are served from different host names or multiple hostnames. That feels unlikely.

Reply:
Another example I am thinking about is that we may move from one cert to another. In that case you may want to ensure one is absent while the other is present.

That said, I'd be ok with a simplification now but I'd ask you to describe that limitation in the type documentation.

Reply:
I think that last example is actually a good one (but I'd rename it to just `truststore`). It would also imply that we purge all other certificates from it. I would suggest a hash instead:
```puppet
truststore { $truststore:
  ensure        => present,
  password_file => $truststore_password_path,
  certificates  => {
      $alias         => $ca_cert,
      $artemis_alias => $client_cert,
  },
}
```

Reply:
Looks very pretty to me :)

One question: the certificates and keys and reality paths to files containing those, right? Not the actual (PEM/DER) encoded data? Otherwise we could be leaking key content to logs etc.

Reply:
While we rarely use it, there is [Sensitive](https://puppet.com/docs/puppet/7.6/lang_data_sensitive.html). It does have some issues. For example, if you use it in a template you must unwrap it and the diff of those files does show the raw value (unless you use `show_diff => false`)

Reply:
> One question: the certificates and keys and reality paths to files containing those, right? Not the actual (PEM/DER) encoded data? Otherwise we could be leaking key content to logs etc.

You are correct. This is just the paths to the files containing the certificates, not the data inside the certificates and keys.

Reply:
Based on the whole discussion, an alternative approach -- https://github.com/theforeman/puppet-certs/pull/334

To which this PR would become targeted at providing the `keystore_certificate` functionality.